core/src/main/java/org/apache/calcite/rex/RexCall.java

```diff
@@ -127,11 +137,83 @@ protected final StringBuilder appendOperands(StringBuilder sb) {
           includeType = RexDigestIncludeType.NO_TYPE;
         }
       }
-      sb.append(((RexLiteral) operand).computeDigest(includeType));
+      operandDigests.add(((RexLiteral) operand).computeDigest(includeType));
+    }
+    int totalLength = (operandDigests.size() - 1) * 2; // commas
+    for (String s : operandDigests) {
+      totalLength += s.length();
+    }
+    sb.ensureCapacity(sb.length() + totalLength);
+    sortOperandsIfNeeded(sb, operands, operandDigests);
+    for (int i = 0; i < operandDigests.size(); i++) {
+      String op = operandDigests.get(i);
+      if (i != 0) {
+        sb.append(", ");
+      }
+      sb.append(op);
     }
     return sb;
   }
 
+  private void sortOperandsIfNeeded(StringBuilder sb,
+      List<RexNode> operands, List<String> operandDigests) {
+    if (operands.isEmpty() || !needNormalize()) {
+      return;
+    }
+    final SqlKind kind = op.getKind();
+    if (SqlKind.SYMMETRICAL_SAME_ARG_TYPE.contains(kind)) {
+      final RelDataType firstType = operands.get(0).getType();
+      for (int i = 1; i < operands.size(); i++) {
+        if (!equalSansNullability(firstType, operands.get(i).getType())) {
+          // Arguments have different type, thus they must not be sorted
```

Comment:
We can make a sort if:
1. There is no decimals in the `PLUS`
2. There is no `INTERVAL`s in `TIMES`

Reply:
Why drop decimals?
Why drop INTERVAL?

Reply:
Because of the return type inference, see `RetirnTypes.NULLABLE_SUM` and `ReturnTypes.PRODUCT_NULLABLE`.

Reply:
I do not follow you.
I ensure the operands have to be of the same type, so I don't see why we should reject reordering in `DECIMAL(5,0) + DECIMAL(5,0)`.

Reply:
Oh, you mis-understood what i mean, i means to support more reordering for `PLUS` and `TIMES` even their operands types are different. For example, `DECIMAL(3,0) + DECIMAL(5,0)` can also be reordered.

Reply:
Do we want to support strict floating-point non-associativity here?
Technically speaking, `PLUS` for FLOATs is not associative :(

https://stackoverflow.com/a/10371890/1261287

>i means to support more reordering for PLUS and TIMES even their operands types are differen

I wanted the first iteration to be as simple as it could be.
Proper cost estimates are more interesting for me than spending time on `DECIMAL(3,0) + DECIMAL(5,0)` normalization.

Reordering of `DECIMAL(3,0) + DECIMAL(5,0)` might be tricky, especially if it is nested: `DECIMAL(3,0) + DECIMAL(3,0) + DECIMAL(5,0)` might probably return different result than `DECIMAL(5,0) + DECIMAL(3,0) + DECIMAL(3,0)`, so I don't want to spend time on exploring those scenarios as they are quite rare for me.

Reply:
I agree for the first iteration, we can keep it as simple as possible.

Why do you think `PLUS` for different data types are rare cases, i believe there would be many cases in production ~

Reply:
I've no idea. We could probably add a logger or counter to so it would count the number of times the normalization was skipped for + / *.

If you have production-like queries/test-cases, it would be nice if you could add them.

However, I still think refining cost estimates and uniqueness metadata (see https://github.com/apache/calcite/pull/1702 ) has better benefits than enhanced + * normalization.

Reply:
In production the types will be the same: https://twitter.com/pysailor_de/status/1213126545642065920 :)